File: src/remi/daemon.py
"""Daemon process — always-on mode for production.

Usage: python -m remi serve

Manages:
- Connector lifecycle (Feishu webhook, etc.)
- Scheduler (heartbeat, memory compaction, reminders)
- PID file (prevent duplicate instances)
- Graceful shutdown (SIGTERM/SIGINT)
"""

from __future__ import annotations

import asyncio
import logging
import os
import signal
import sys
from pathlib import Path

from remi.config import RemiConfig, load_config
from remi.core import Remi
from remi.providers.claude_cli import ClaudeCLIProvider
from remi.scheduler.jobs import Scheduler

logger = logging.getLogger(__name__)


class RemiDaemon:
    """Always-on daemon process."""

    def __init__(self, config: RemiConfig | None = None) -> None:
        self.config = config or load_config()
        self._shutdown_event = asyncio.Event()

    # ── PID file management ──────────────────────────────────

    def _write_pid(self) -> None:
        self.config.pid_file.parent.mkdir(parents=True, exist_ok=True)
        self.config.pid_file.write_text(str(os.getpid()))
        logger.info("PID file written: %s (pid=%d)", self.config.pid_file, os.getpid())

    def _remove_pid(self) -> None:
        if self.config.pid_file.exists():
            self.config.pid_file.unlink()

    def _check_existing(self) -> None:
        if not self.config.pid_file.exists():
            return
        try:
            pid = int(self.config.pid_file.read_text().strip())
            os.kill(pid, 0)  # Check if process exists
            print(f"Remi daemon already running (pid={pid}). Exiting.", file=sys.stderr)
            sys.exit(1)
        except (ProcessLookupError, ValueError):
            # Stale PID file — remove it
            self._remove_pid()

    # ── Signal handling ──────────────────────────────────────

    def _setup_signals(self) -> None:
        loop = asyncio.get_event_loop()
        for sig in (signal.SIGTERM, signal.SIGINT):
            loop.add_signal_handler(sig, self._handle_signal, sig)

    def _handle_signal(self, sig: signal.Signals) -> None:
        logger.info("Received %s, shutting down...", sig.name)
        self._shutdown_event.set()

    # ── Build components ─────────────────────────────────────

    def _build_remi(self) -> Remi:
        remi = Remi(self.config)

        # Register providers
        provider = self._build_provider()
        remi.add_provider(provider)

        # Register memory tools on provider (if supported)
        self._register_memory_tools(provider, remi)

        # Register fallback if configured
        if self.config.provider.fallback:
            try:
                fallback = self._build_provider(self.config.provider.fallback)
                remi.add_provider(fallback)
            except Exception as e:
                logger.warning("Failed to build fallback provider: %s", e)

        return remi

    def _register_memory_tools(self, provider, remi: Remi) -> None:
        """Register memory tools on a provider that supports tool registration."""
        register = getattr(provider, "register_tools_from_dict", None)
        if not register:
            return

        try:
            from remi.tools.memory_tools import get_memory_tools

            tools = get_memory_tools(remi.memory)
            register(tools)
            logger.info("Registered %d memory tools on %s", len(tools), provider.name)
        except Exception as e:
            logger.warning("Failed to register memory tools: %s", e)

    def _build_provider(self, name: str | None = None):
        name = name or self.config.provider.name
        if name == "claude_cli":
            return ClaudeCLIProvider(
                model=self.config.provider.model,
                timeout=self.config.provider.timeout,
                allowed_tools=self.config.provider.allowed_tools,
            )
        else:
            raise ValueError(f"Unknown provider: {name}")

    def _build_connectors(self, remi: Remi) -> None:
        if self.config.feishu.app_id:
            try:
                from remi.connectors.feishu import FeishuConnector

                connector = FeishuConnector(self.config.feishu)
                remi.add_connector(connector)
            except ImportError:
                logger.warning("Feishu connector unavailable (install 'remi[feishu]')")

    # ── Main run loop ────────────────────────────────────────

    async def run(self) -> None:
        self._check_existing()
        self._write_pid()
        self._setup_signals()

        remi = self._build_remi()
        self._build_connectors(remi)

        scheduler = Scheduler(remi, self.config)

        logger.info("Remi daemon starting (provider=%s)", self.config.provider.name)

        try:
            await asyncio.gather(
                remi.start(),
                scheduler.start(self._shutdown_event),
            )
        except asyncio.CancelledError:
            pass
        finally:
            await remi.stop()
            self._remove_pid()
            logger.info("Remi daemon stopped.")

File: src/daemon.ts
/**
 * Daemon process — always-on mode for production.
 *
 * Usage: bun run src/main.ts serve
 *
 * Manages:
 * - Connector lifecycle
 * - Scheduler (heartbeat, memory compaction, reminders)
 * - PID file (prevent duplicate instances)
 * - Graceful shutdown (SIGTERM/SIGINT)
 */

import { existsSync, readFileSync, writeFileSync, unlinkSync, mkdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { homedir } from "node:os";
import { spawn } from "node:child_process";
import type { Connector } from "./connectors/base.js";
import type { RemiConfig } from "./config.js";
import { loadConfig, migrateConfigFile } from "./config.js";
import { Remi } from "./core.js";
import { ClaudeCLIProvider } from "./providers/claude-cli/index.js";
import { CronTimer } from "./scheduler/cron-timer.js";
import { FeishuConnector } from "./connectors/feishu/index.js";
import { flushDedupCacheSync } from "./connectors/feishu/receive.js";
import { AuthStore, FeishuAuthAdapter, ByteDanceSSOAdapter } from "./auth/index.js";
import type { TokenSyncRule } from "./auth/token-sync.js";
import { createLogger } from "./logger.js";
import { writeEcosystem, runBuildsSync, getEcosystemPath } from "./pm2.js";

const log = createLogger("daemon");

export class RemiDaemon {
  config: RemiConfig;
  private _abortController = new AbortController();

  constructor(config?: RemiConfig) {
    this.config = config ?? loadConfig();
  }

  // ── PID file management ──────────────────────────────────

  private _writePid(): void {
    const dir = dirname(this.config.pidFile);
    if (!existsSync(dir)) {
      mkdirSync(dir, { recursive: true });
    }
    writeFileSync(this.config.pidFile, String(process.pid));
    log.info(`PID file written: ${this.config.pidFile} (pid=${process.pid})`);
  }

  private _removePid(): void {
    if (existsSync(this.config.pidFile)) {
      unlinkSync(this.config.pidFile);
    }
  }

  private _checkExisting(): void {
    if (!existsSync(this.config.pidFile)) return;

    let pid: number;
    try {
      pid = parseInt(readFileSync(this.config.pidFile, "utf-8").trim(), 10);
    } catch {
      this._removePid();
      return;
    }
    if (isNaN(pid)) {
      this._removePid();
      return;
    }

    // Check if process is still running
    let alive = false;
    try {
      process.kill(pid, 0);
      alive = true;
    } catch {
      // Process doesn't exist — stale PID file
      this._removePid();
      return;
    }

    if (!alive) return;

    // Takeover: send SIGTERM and wait for old process to exit
    log.info(`Found existing Remi daemon (pid=${pid}), sending SIGTERM to take over...`);
    try { process.kill(pid, "SIGTERM"); } catch { /* already gone */ }

    const deadline = Date.now() + 10_000; // 10s timeout
    while (Date.now() < deadline) {
      try {
        process.kill(pid, 0);
      } catch {
        // Old process exited
        log.info(`Old daemon (pid=${pid}) exited.`);
        this._removePid();
        return;
      }
      Bun.sleepSync(200);
    }

    // Still alive after timeout — force kill
    log.warn(`Old daemon (pid=${pid}) did not exit in 10s, sending SIGKILL...`);
    try { process.kill(pid, "SIGKILL"); } catch { /* ignore */ }
    Bun.sleepSync(500);
    this._removePid();
  }

  // ── Signal handling ──────────────────────────────────────

  private _setupSignals(): void {
    const handler = (sig: string) => {
      log.info(`Received ${sig}, shutting down...`);
      this._abortController.abort();
    };
    process.on("SIGTERM", () => handler("SIGTERM"));
    process.on("SIGINT", () => handler("SIGINT"));
  }

  // ── Build components ─────────────────────────────────────

  _buildRemi(): Remi {
    const remi = new Remi(this.config);

    // 1. Initialize AuthStore (1Passport) with token sync rules
    const syncRules: TokenSyncRule[] | undefined =
      this.config.tokenSync.length > 0
        ? (this.config.tokenSync as TokenSyncRule[])
        : undefined; // undefined → use defaults
    const authStore = new AuthStore(join(homedir(), ".remi", "auth"), syncRules);
    const hasFeishuCreds = !!(this.config.feishu.appId && this.config.feishu.appSecret);
    if (hasFeishuCreds) {
      authStore.registerAdapter(
        new FeishuAuthAdapter({
          appId: this.config.feishu.appId,
          appSecret: this.config.feishu.appSecret,
          domain: this.config.feishu.domain,
          userAccessToken: this.config.feishu.userAccessToken || undefined,
        }),
      );
    }
    if (this.config.bytedanceSso?.clientId) {
      authStore.registerAdapter(
        new ByteDanceSSOAdapter(this.config.bytedanceSso),
      );
      log.info("Registered ByteDance SSO adapter (1Passport)");
    }
    remi.authStore = authStore;

    // 2. Register providers
    const provider = this._buildProvider();
    remi.addProvider(provider);

    // 3. Feishu document tools (lark_fetch/lark_render/lark_search/lark_auth) are now
    // provided via MCP server (bytedance.lark_parser built-in lark-mcp-server).
    // Memory tools (recall/remember) via MCP server (src/mcp/memory-server.ts).

    // Register fallback if configured
    if (this.config.provider.fallback) {
      try {
        const fallback = this._buildProvider(this.config.provider.fallback);
        remi.addProvider(fallback);
      } catch (e) {
        log.warn("Failed to build fallback provider:", e);
      }
    }

    // 4. Register Feishu connector with token provider
    if (hasFeishuCreds) {
      // Merge bot profile groups into allowedGroups only (not monitorGroups —
      // bot profile groups require @bot or @triggerUser to respond)
      const feishuConfig = { ...this.config.feishu };
      for (const bot of this.config.bots) {
        for (const g of bot.groups) {
          if (!feishuConfig.allowedGroups.includes(g)) {
            feishuConfig.allowedGroups = [...feishuConfig.allowedGroups, g];
          }
        }
      }

      const feishu = new FeishuConnector(feishuConfig);
      feishu.setTokenProvider(() => authStore.getToken("feishu", "tenant"));
      feishu.setBotProfiles(this.config.bots);
      remi.addConnector(feishu);
      log.info(`Registered Feishu connector (with 1Passport, ${this.config.bots.length} bot profiles)`);
    }

    // Register restart handler (only triggered by /restart slash command)
    remi.onRestart((info) => this._restart(info));

    return remi;
  }

  private _buildProvider(name?: string | null) {
    const n = name ?? this.config.provider.name;
    if (n === "claude_cli") {
      return new ClaudeCLIProvider({
        model: this.config.provider.model,
        timeout: this.config.provider.timeout,
        allowedTools: this.config.provider.allowedTools,
        cwd: homedir(),
      });
    }
    throw new Error(`Unknown provider: ${n}`);
  }

  // ── Restart notification ────────────────────────────────

  private get _restartNotifyPath(): string {
    return join(homedir(), ".remi", "restart-notify.json");
  }

  private _saveRestartNotify(info: { chatId: string; connectorName?: string }): void {
    const dir = join(homedir(), ".remi");
    if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
    writeFileSync(this._restartNotifyPath, JSON.stringify(info));
  }

  /** After startup, check if we need to notify someone that restart succeeded. */
  private async _sendRestartNotify(remi: Remi): Promise<void> {
    const filePath = this._restartNotifyPath;
    if (!existsSync(filePath)) return;

    let info: { chatId: string; connectorName?: string };
    try {
      const raw = readFileSync(filePath, "utf-8");
      info = JSON.parse(raw);
      unlinkSync(filePath);
      log.info(`Restart notify: connector=${info.connectorName}, chatId=${info.chatId}`);
    } catch (e) {
      log.warn("Restart notify: failed to read file:", e);
      if (existsSync(filePath)) unlinkSync(filePath);
      return;
    }

    // Find the matching connector
    const connectors = remi["_connectors"] as Connector[];
    const connector = connectors.find(
      (c) => c.name === (info.connectorName ?? ""),
    );
    if (!connector) {
      log.warn(
        `Restart notify: connector "${info.connectorName}" not found (available: ${connectors.map((c) => c.name).join(", ")})`,
      );
      return;
    }

    // Retry — Feishu API may need time to obtain access token on cold start
    const maxRetries = 3;
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        await connector.reply(info.chatId, { text: "Remi 重启成功，已上线。" });
        log.info(`Restart notification sent to ${info.connectorName}:${info.chatId}`);
        return;
      } catch (e) {
        log.warn(`Restart notify attempt ${attempt}/${maxRetries} failed: ${String(e)}`);
        if (attempt < maxRetries) {
          await new Promise((r) => setTimeout(r, 3000));
        }
      }
    }
    log.error("Restart notification failed after all retries.");
  }

  // ── Self-restart ────────────────────────────────────────

  private _restart(info: { chatId: string; connectorName?: string }): void {
    log.info("Restart requested — rebuilding services and triggering PM2 restart...");

    this._saveRestartNotify(info);
    flushDedupCacheSync();

    // Run build steps for services that need them
    runBuildsSync(this.config);

    // Regenerate ecosystem config (picks up any config changes)
    writeEcosystem(this.config);

    // Trigger PM2 restart (detached so it outlives our process)
    const child = spawn("pm2", ["restart", getEcosystemPath(), "--update-env"], {
      detached: true,
      stdio: "ignore",
    });
    child.unref();

    // PM2 will send SIGTERM which triggers our graceful shutdown
  }

  // ── Main run loop ────────────────────────────────────────

  async run(): Promise<void> {
    this._checkExisting();
    this._writePid();
    this._setupSignals();

    // One-time migration: [scheduler] + [[scheduled_skills]] → [[cron.jobs]]
    if (migrateConfigFile()) {
      log.info("Config migrated: [scheduler] + [[scheduled_skills]] → [[cron.jobs]]");
      this.config = loadConfig();
    }

    const remi = this._buildRemi();
    const cronTimer = new CronTimer(remi, this.config);

    log.info("=".repeat(60));
    log.info(`Remi daemon starting at ${new Date().toISOString()} (pid=${process.pid}, provider=${this.config.provider.name})`);

    // Send restart notification after connectors have time to fully initialize
    setTimeout(() => this._sendRestartNotify(remi), 5000);

    try {
      await Promise.all([
        remi.start(),
        cronTimer.start(this._abortController.signal),
      ]);
    } catch (e) {
      if ((e as Error).name !== "AbortError") {
        // Abort CronTimer if remi.start() or any component fails
        this._abortController.abort();
        throw e;
      }
    } finally {
      this._abortController.abort(); // Ensure CronTimer stops in all exit paths
      await remi.stop();
      this._removePid();
      log.info("Remi daemon stopped.");
    }
  }
}
